pkg/resources/provisioning.cattle.io/v1/cluster/validator.go

```diff
@@ -194,11 +215,78 @@ func (p *provisioningAdmitter) validateDataDirectories(request *admission.Reques
 	if response := p.validateSystemAgentDataDirectory(oldCluster, newCluster); !response.Allowed {
 		return response
 	}
-	if oldCluster.Spec.RKEConfig.DataDirectories.Provisioning != newCluster.Spec.RKEConfig.DataDirectories.Provisioning {
+	if oldCluster.Spec.RKEConfig.DataDirectories.K8sDistro != distro {
+		return admission.ResponseBadRequest("Distro data directory cannot be changed after cluster creation")
+	}
+	if oldCluster.Spec.RKEConfig.DataDirectories.Provisioning != provisioning {
 		return admission.ResponseBadRequest("Provisioning data directory cannot be changed after cluster creation")
 	}
-	if oldCluster.Spec.RKEConfig.DataDirectories.K8sDistro != newCluster.Spec.RKEConfig.DataDirectories.K8sDistro {
-		return admission.ResponseBadRequest("Distro data directory cannot be changed after cluster creation")
+
+	return admission.ResponseAllowed()
+}
+
+// validateDataDirectoryFormat ensures that no data directory contains a relative path, environment variables,
+// shell expressions, or references to the current or parent directory via use of "./" and "../" respectively.
+// dir is the path of the data directory, and name corresponds to a print friendly name for this data directory.
+func validateDataDirectoryFormat(dir, name string) *admissionv1.AdmissionResponse {
+	if dir == "" {
+		return admission.ResponseAllowed()
+	}
+	if !filepath.IsAbs(dir) {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory must be an absolute path", name))
+	}
+	if strings.ContainsAny(dir, "${}()`") {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory cannot contain shell expressions", name))
+	}
+	if filepath.Clean(dir) != dir {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory is not clean", name))
+	}
+
+	return admission.ResponseAllowed()
+}
+
+// validateDataDirectoryHierarchy ensures that no directories are equal, and no directories include other directories.
+// dataDirs is a map with keys corresponding to print friendly names for these data directories, and values representing
+// the specific data directories.
+func validateDataDirectoryHierarchy(dataDirs map[string]string) *admissionv1.AdmissionResponse {
+	paths := make([]string, 0, len(dataDirs))
+	for _, dir := range dataDirs {
+		// do not attempt to validate empty directory
+		if dir == "" {
+			continue
+		}
+		paths = append(paths, dir)
+	}
+
+	for i := range paths {
+		for j := i + 1; j < len(paths); j++ {
+			path1 := paths[i]
+			path2 := paths[j]
+
+			if path1 == path2 {
+				return admission.ResponseBadRequest(
+					fmt.Sprintf("%s data directory cannot be equal to %s data directory", dataDirs[path1], dataDirs[path2]))
```

Comment:
The keys in the map `dataDirs` are "Distro", "Provisioning" and "System Agent", not the path, which will result in the empty string being placed in the produced message.

Reply:
good catch!

Reply:
Good catch, I had it the other way around but flipped them because if multiple keys were identical it would overwrite before validating, I'll update.